### PR TITLE
Add schema for field indexing

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
See go/firestore-composite-index

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java [hunk 1]
```diff
@@ -336,24 +336,32 @@ private void createV1RemoteDocumentCache() {
...
-  private void createLocalDocumentsCollectionIndex() {
+  private void createFieldIndex() {
+    // Create two tables to support document indexing. `index_configuration` holds the
+    // configuration for all currently active indices. All entries apply for all users and it is
+    // not possible to only enable indices for a subset of users. `field_index` holds the index
+    // values. If there are pending mutations that affect an indexed field, an additional index
+    // entry is created per mutated field.
     ifTablesDontExist(
-        new String[] {"collection_index"},
+        new String[] {"index_configuration", "field_index"},
```
Comment (wu-hui) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:346]: Looks like other tables use plural.

Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:346]: And what about `field_index_entries`, too long?

**Review by wu-hui — COMMENTED** [review]
Please add a simple test to verify tables are created.

Also, not related to code review..when we are sure the schemas are not going to change, we can just release it to the users, is that the plan?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java [hunk 2]
```diff
@@ -336,24 +336,32 @@ private void createV1RemoteDocumentCache() {
         });
   }
 
-  // TODO(indexing): Put the schema version in this method name.
-  private void createLocalDocumentsCollectionIndex() {
+  private void createFieldIndex() {
+    // Create two tables to support document indexing. `index_configuration` holds the
+    // configuration for all currently active indices. All entries apply for all users and it is
+    // not possible to only enable indices for a subset of users. `field_index` holds the index
+    // values. If there are pending mutations that affect an indexed field, an additional index
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:343]: I think this should just be "an entry", can you double check?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:346]: I don't have a strong opinion. I made this "index_entries", which is shorter but still contains your suggestion. It also uses the same prefix as "index_configuration", which can be nice for grouping.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java [hunk 3]
```diff
@@ -176,7 +176,7 @@ void runMigrations(int fromVersion, int toVersion) {
 
     if (fromVersion < INDEXING_SUPPORT_VERSION && toVersion >= INDEXING_SUPPORT_VERSION) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:177]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java [hunk 4]
```diff
@@ -336,24 +336,32 @@ private void createV1RemoteDocumentCache() {
         });
   }
 
-  // TODO(indexing): Put the schema version in this method name.
-  private void createLocalDocumentsCollectionIndex() {
+  private void createFieldIndex() {
+    // Create two tables to support document indexing. `index_configuration` holds the
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:340]: Moved to method comment, expanded it a bit and improved formatting.

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks for the review. I also added a very basic test.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java [hunk 5]
```diff
@@ -336,24 +336,37 @@ private void createV1RemoteDocumentCache() {
...
-  // TODO(indexing): Put the schema version in this method name.
-  private void createLocalDocumentsCollectionIndex() {
+  /**
+   * Creates the necessary tables to support document indexing.
+   *
+   * <p>The `index_configuration` table holds the configuration for all indices. Entries in this
+   * table apply for all users. It is not possible to only enable indices for a subset of users.
+   *
+   * <p>The `index_entries` table holds the index values themselves. An index value is created for
+   * every indexed field. If there are pending mutations that affect an indexed field, an additional
```
Comment (wu-hui) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:346]: This is not technically true is it? This suggests if there is a composite index with 2 fields, there will be two entries in this table.

How about `An entry is created for each value of the indexed field(s)...`

**Review by wu-hui — APPROVED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java (hunk 5 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java:346]: You are correct. I changed it to say " An index value is created for each field combination that matches a configured index."

Thanks for catching.

**Pull request merged**